### PR TITLE
Discriminator support for oneOf schema

### PR DESCRIPTION
**Pull request by dps123**
Signed-off-by: Pavel Demenkov demps@inbox.ru

### CODE REVIEW

File: src/test/resources/tck/openapi3/discriminator.json [hunk 1]
```diff
@@ -74,5 +74,174 @@
...
+        },
+        "valid": true
+      },
+      {
+        "description": "should be invalid with id A and property value_b",
+        "data": {
+          "id": "A",
+          "value_b": "b"
+        },
+        "valid": false
```
Comment (slinkydeveloper) [inline src/test/resources/tck/openapi3/discriminator.json:134]: Why this is invalid? This should be valid because there isn't any `additionalProperties: false` in the A schema, so `value_b` should be catched as additional property

File: src/test/resources/tck/openapi3/discriminator.json [hunk 2]
```diff
@@ -74,5 +74,174 @@
...
+      },
+      {
+        "description": "should be invalid without id",
+        "data": {
+          "value_b": "b"
+        },
+        "valid": false
+      },
+      {
+        "description": "should be invalid with id A and without value_a",
```
Comment (slinkydeveloper) [inline src/test/resources/tck/openapi3/discriminator.json:152]: Can you add 2 test cases: one with the empty data (`"data": {}`) and one with only the id `B`

Comment (slinkydeveloper) [reply src/test/resources/tck/openapi3/discriminator.json:152]: Same below

File: src/test/resources/tck/openapi3/discriminator.json (hunk 1 above)
Comment (slinkydeveloper) [reply src/test/resources/tck/openapi3/discriminator.json:134]: Same question for the same test case below

File: src/main/java/io/vertx/ext/json/schema/common/OneOfValidatorFactory.java [hunk 3]
```diff
@@ -11,16 +11,34 @@
...
 import io.vertx.ext.json.schema.NoSyncValidationException;
 import io.vertx.ext.json.schema.ValidationException;
 
 import java.util.Arrays;
 import java.util.stream.Collectors;
 
 import static io.vertx.ext.json.schema.ValidationException.createException;
+import java.util.Map;
 
 public class OneOfValidatorFactory extends BaseCombinatorsValidatorFactory {
```
Comment (slinkydeveloper) [inline src/main/java/io/vertx/ext/json/schema/common/OneOfValidatorFactory.java:25]: This code to support discriminator shouldn't be in common, but in the openapi3 package because it's related only to openapi 3

**Review by slinkydeveloper — COMMENTED** [review]
I did a first pass, I'll properly review it in the next days

File: src/test/resources/tck/openapi3/discriminator.json (hunk 1 above)
Comment (dps123) [reply src/test/resources/tck/openapi3/discriminator.json:134]: It is because `"required": ["id", "value_a"]` in the A schema

File: src/main/java/io/vertx/ext/json/schema/openapi3/OneOfDiscriminatorValidatorFactory.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+      if (isSync()) return validateSyncAsAsync(context, in);
+      List<Future<Void>> futures = Arrays.stream(schemas).filter(s -> checkDiscriminator(s, in)).map(s -> s.validateAsync(context, in)).collect(Collectors.toList());
+      if (futures.isEmpty()) {
+        return Future.failedFuture(createException("No schema matches", "oneOf", in));
+      } else {
+        return FutureUtils.oneOf(futures).recover(err -> Future.failedFuture(createException("No schema matches", "oneOf", in, err)));
+      }
+    }
+
+    private boolean checkDiscriminator(SchemaInternal schema, Object in) {
```
Comment (slinkydeveloper) [inline src/main/java/io/vertx/ext/json/schema/openapi3/OneOfDiscriminatorValidatorFactory.java:91]: I'm wondering, can we place some of this code that parses the original schema in the factory/initialization code more than on the hot path?

You can also reorganize the data structure of the validator to have a map discriminator -> schema. Note that you don't have to use the same validator both for oneOf case and oneOf + discriminator case. Feel free to reach me if you have more questions

Comment (slinkydeveloper) [reply src/main/java/io/vertx/ext/json/schema/openapi3/OneOfDiscriminatorValidatorFactory.java:91]: ping @dps123

**Comment by slinkydeveloper** [comment]
Links to #31

File: src/main/java/io/vertx/ext/json/schema/openapi3/OneOfDiscriminatorValidatorFactory.java (hunk 4 above)
Comment (dps123) [reply src/main/java/io/vertx/ext/json/schema/openapi3/OneOfDiscriminatorValidatorFactory.java]: @slinkydeveloper Moved the schema parsing to the validator initializer

Comment (slinkydeveloper) [reply src/main/java/io/vertx/ext/json/schema/openapi3/OneOfDiscriminatorValidatorFactory.java]: Can you please check my comments on https://github.com/eclipse-vertx/vertx-json-schema/issues/31?

**Review by murakobardenas — APPROVED** [review]

**Comment by pk-work** [comment]
We have deprecated this API and the new one does not uses this code.
